### PR TITLE
Added basic hook for parameter changes

### PR DESCRIPTION
**Pull request by rohbotics**
See #243

### CODE REVIEW

**Comment by wjwwood** [comment]
lgtm, we should run CI over it before merging though.

**Comment by wjwwood** [comment]
Here are some CI jobs for your pr:

- Linux:
  - [Build Status](http://ci.ros2.org/job/ci_linux/1675/)
- OS X:
  - [Build Status](http://ci.ros2.org/job/ci_osx/1246/)

**Comment by rohbotics** [comment]
The only failing test is from this issue:
https://github.com/osrf/osrf_pycommon/issues/25

File: rclcpp/include/rclcpp/node.hpp [hunk 1]
```diff
@@ -353,6 +353,9 @@ class Node : public std::enable_shared_from_this<Node>
   size_t
   count_graph_users();
 
+  template<typename CallbackT>
+  void on_param_change(CallbackT && callback);
```
Comment (dirk-thomas) [inline rclcpp/include/rclcpp/node.hpp]: I would expect an actual callback function to have such a name (`on_param_change`). For this I think `register_param_change_callback` (or similar) would be better.

Also from the name itself it is not clear that repeated invocations will overwrite previous callback functions. That should at least be documented in a docblock. Maybe there should also be a way to unregister a callback again?

**Comment by rohbotics** [comment]
@dirk-thomas Updated with your suggestions

**Comment by dirk-thomas** [comment]
Thanks, lgtm.

**Comment by wjwwood** [comment]
I think I fixed the `osrf_pycommon` issue. I started new CI since some changes were made here as well:

- Linux:
  - [Build Status](http://ci.ros2.org/job/ci_linux/1677/)
- OS X:
  - [Build Status](http://ci.ros2.org/job/ci_osx/1247/)

**Comment by wjwwood** [comment]
lgtm

**Pull request merged**